flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/RowTimeDeduplicateKeepFirstRowFunction.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.runtime.typeutils.InternalTypeInfo;
+import org.apache.flink.util.Collector;
+
+import static org.apache.flink.table.runtime.operators.deduplicate.utils.DeduplicateFunctionHelper.checkInsertOnly;
+import static org.apache.flink.table.runtime.operators.deduplicate.utils.DeduplicateFunctionHelper.shouldKeepCurrentRow;
+import static org.apache.flink.table.runtime.util.StateConfigUtil.createTtlConfig;
+
+/**
+ * This function is used to deduplicate on keys and keeps only first row on row time. It produces
+ * append only stream thanks to emitting results only via firing the timers.
+ */
+public class RowTimeDeduplicateKeepFirstRowFunction
+        extends KeyedProcessFunction<RowData, RowData, RowData> {
+
+    private static final long serialVersionUID = 1L;
+
+    // the TypeInformation of the values in the state.
+    private final TypeInformation<RowData> typeInfo;
+    private final long stateRetentionTime;
+    private final int rowtimeIndex;
+
+    // state stores previous message under the key.
+    protected ValueState<RowData> waitingToEmitOnTimerState;
+    protected ValueState<Boolean> alreadyEmittedState;
+
+    public RowTimeDeduplicateKeepFirstRowFunction(
+            InternalTypeInfo<RowData> typeInfo, long minRetentionTime, int rowtimeIndex) {
+        this.typeInfo = typeInfo;
+        this.stateRetentionTime = minRetentionTime;
+        this.rowtimeIndex = rowtimeIndex;
+    }
+
+    @Override
+    public void open(OpenContext openContext) throws Exception {
+        super.open(openContext);
+
+        // We don't enable TTL on the timer's state, because we rely on the state cleaning up on
+        // watermark. Also otherwise TTL clean up before firing the watermark would cause a data
+        // loss.
+        ValueStateDescriptor<RowData> timerStateDesc =
+                new ValueStateDescriptor<>("waiting-to-emit-on-timer", typeInfo);
+        waitingToEmitOnTimerState = getRuntimeContext().getState(timerStateDesc);
+
+        ValueStateDescriptor<Boolean> stateDesc =
+                new ValueStateDescriptor<>("already-emitted-state-boolean", Types.BOOLEAN);
+        StateTtlConfig ttlConfig = createTtlConfig(stateRetentionTime);
+        if (ttlConfig.isEnabled()) {
+            stateDesc.enableTimeToLive(ttlConfig);
+        }
+        alreadyEmittedState = getRuntimeContext().getState(stateDesc);
+    }
+
+    @Override
+    public void processElement(RowData input, Context ctx, Collector<RowData> out)
+            throws Exception {
+        checkInsertOnly(input);
+        Boolean allreadyEmitted = alreadyEmittedState.value();
+        if (allreadyEmitted != null && allreadyEmitted) {
+            // result has already been emitted, we can not retract/emit anything different.
+            return;
+        }
+        long rowtime = input.getLong(rowtimeIndex);
+        if (rowtime < ctx.timerService().currentWatermark()) {
```

Comment:
What about introducing a metric to log late num used for debugging just like other window operators?